File: java/src/com/edmundmartin/algos/stringsArrays/PalindromePermutation.java
package com.edmundmartin.algos.stringsArrays;

public class PalindromePermutation {

    private int getCharNumber(Character c) {
        int a = Character.getNumericValue('a');
        int z = Character.getNumericValue('z');
        int val = Character.getNumericValue(c);
        if (a <= val && val <= z) {
            return val - a;
        }
        return -1;
    }

    private int[] buildCharFrequencyTable(String phrase) {
        int[] table = new int[Character.getNumericValue('z') - Character.getNumericValue('a') - 1];

        for (char c: phrase.toCharArray()) {
            int x = getCharNumber(c);
            if (x != -1) {
                table[x]++;
            }
        }
        return table;
    }

    /* Check no more than one character has na odd count */
    private boolean checkMaxOneOdd(int[] table) {
        boolean foundOdd = false;
        for (int count: table) {
            if (count % 2 == 1) {
                if (foundOdd) {
                    return false;
                }
                foundOdd = true;
            }
        }
        return true;
    }

    public boolean isPermutationPalindromeNaive(String phrase) {
        int[] table = buildCharFrequencyTable(phrase);
        return checkMaxOneOdd(table);
    }

    /* Does not improve big O - but is faster */
    public boolean isPermutationPalindromeOptimised(String phrase) {
        int countOdd = 0;
        int[] table = new int[Character.getNumericValue('z') - Character.getNumericValue('a') + 1];
        for (char c: phrase.toCharArray()) {
            int x = getCharNumber(c);
            if (x != -1) {
                table[x]++;
                if (table[x] % 2 == 1) {
                    countOdd++;
                } else {
                    countOdd--;
                }
            }
        }
        return countOdd <= 1;
    }
}

File: golang/pkg/strings_array/palindrome_permutation.go
package strings_array


func buildFreqTable(input string) map[rune]int {
	freqTable := make(map[rune]int)
	for _, r := range input {
		freqTable[r]++
	}
	return freqTable
}

func checkMaxOneOdd(freqTable map[rune]int) bool {
	foundOdd := false
	for _, value := range freqTable {
		if value % 2 == 1 {
			if foundOdd {
				return false
			}
			foundOdd = true
		}
	}
	return true
}

func IsPermutationPalindromeNaive(phrase string) bool {
	freqTable := buildFreqTable(phrase)
	return checkMaxOneOdd(freqTable)
}

func IsPermutationPalindromeOptimised(phrase string) bool {
	countOdd := 0
	freqTable := make(map[rune]int)
	for _, r := range phrase {
		freqTable[r]++
		val, _ := freqTable[r]
		if val % 2 == 1 {
			countOdd++
		} else {
			countOdd--
		}
	}
	return countOdd <= 1
}